### ISSUE TITLE
REDIS configuration causing error on "admin/utilities" but not "admin/utilities/php-info"

### ISSUE THREAD

**Issue by tonyclemmey** [craftcms/cms#7161]
### Description

I am trying to configure sessions to use REDIS with a load-balanced cluster setup but getting service unavailable/internal server errors/ 500 error breaking a specific page such as "admin/utilities/" but not "admin/utilities/php-info" or other admin areas for example.

Links I have been following:

https://craftcms.com/docs/3.x/config/#application-configuration.
https://github.com/craftcms/cms/issues/3904
https://github.com/craftcms/cms/issues/6796
https://craftcms.stackexchange.com/questions/25550/craft-cms-3-redis-sessions-cause-error
https://nystudio107.com/blog/the-case-of-the-missing-php-session

A test file has been placed on the server cluster to ensure REDIS is working. Using Port 6389, on 127.0.0.1.

```
<?php
ini_set('display_errors', 'on');
define('REDIS_HOST', getenv('REDIS_HOST') ?: '127.0.0.1');
define('REDIS_PORT', getenv('REDIS_PORT') ?: '6389');
ini_set('session.save_handler', 'redis');
ini_set('session.save_path', 'tcp://'.REDIS_HOST.':'.REDIS_PORT);
session_start();
phpinfo();
?>
```

Php info will show the local session save path = REDIS. If I then run `# redis-cli -p 6389 monitor` I see the output when access test file.

On the Craft site in `config/app.php` I have the added the REDIS component, for the session.  ( and composer require --prefer-dist yiisoft/yii2-redis is run before code is deployed)

```
        'components' => [
            'redis' => [
                'class' => yii\redis\Connection::class,
                // 'hostname' => getenv('REDIS_HOST'),
                // 'port' => getenv('REDIS_PORT'),
                'hostname' => '127.0.0.1',
                'port' => 6389,
                'database' => 0,
            ],
            'cache' => [
                'class' => yii\redis\Cache::class,
                'redis' => [
                	// 'hostname' => getenv('REDIS_HOST'),
                	// 'port' => getenv('REDIS_PORT'),
                    'hostname' => '127.0.0.1',
                    'port' => 6379,
                    'database' => 1,
                ],
            ],
            //'session' => function() {
                // Get the default component config
                //$config = craft\helpers\App::sessionConfig();

                // Override the class to use Redis' session class
                //$config['class'] = yii\redis\Session::class;

                // Instantiate and return it
                //return Craft::createObject($config);
            //},
            'session' => [
                'class' => \yii\redis\Session::class,
                 'redis' => [
	                // 'hostname' => getenv('REDIS_HOST'),
	                // 'port' => getenv('REDIS_PORT'),
	                'hostname' => '127.0.0.1',
	                'port' => 6389,
	                'database' => 0,
                ],
                'as session' => [
                    'class' => \craft\behaviors\SessionBehavior::class,
                ],
            ],
        ],
```

In DOTenv file I have added the REDIS values anyway.

```
# Redis
REDIS_HOST="127.0.0.1"
REDIS_PORT=6389
```

In the `web/index.php` I have added the following. (Same as what is in test file)

```
// Load Redis
ini_set('display_errors', 'on');
define('REDIS_HOST', getenv('REDIS_HOST') ?: '127.0.0.1');
define('REDIS_PORT', getenv('REDIS_PORT') ?: '6389');
ini_set('session.save_handler', 'redis');
ini_set('session.save_path', 'tcp://'.REDIS_HOST.':'.REDIS_PORT);
session_start();
```

When I run craft I get an internal server error breaking on "admin/utilities" but not "admin/utilities/php-info" or "admin/utilities/logs". 

If I run `# redis-cli -p 6389 monitor` no output is displayed. 

When I turn on DEV mode I get warnings about ini_set(): A session is active. You cannot change the session module's ini settings at this time. New browser window/private mode etc

### Steps to reproduce

1. Add REDIS to `config/app.php` 
2. 

### Additional info

- Craft version: 3.5.8
- PHP version: 7.4.8
- Database driver & version: MariaDB 10.3
- Plugins & versions:
- Redis version: 5.3.1

**Comment by brandonkelly** [comment]
What is the actual error message? Easiest way to find it is by following these steps:

1. Clear out your `storage/logs/` folder
2. Reproduce the error
3. Open `storage/logs/web.log` and search for `[error]`

**Comment by tonyclemmey** [comment]
Done.

The error looks to be related to the REDIS configuration and keeps saying "ini_set(): A session is active. You cannot change the session module's ini settings at this time" If I remove the REDIS configuration the no issues.

web.log

```
2020-11-23 12:03:54 - error - yii\base\ErrorException:2
yii\base\ErrorException: ini_set(): A session is active. You cannot change the session module's ini settings at this time in /home/webmaster/web/live/public_html/web/index.php:28
Stack trace:
#0 /home/webmaster/web/live/public_html/vendor/craftcms/cms/src/web/ErrorHandler.php(76): yii\base\ErrorHandler->handleError(2, 'ini_set(): A se...', '/home/webmaster...', 28)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'ini_set(): A se...', '/home/webmaster...', 28, Array)
#2 /home/webmaster/web/live/public_html/web/index.php(28): ini_set('session.save_ha...', 'redis')
#3 {main}
```

phperrors.log

```
[23-Nov-2020 12:03:54 America/Los_Angeles] An Error occurred while handling another error:
yii\web\HeadersAlreadySentException: Headers already sent. in /home/webmaster/web/live/public_html/vendor/yiisoft/yii2/web/Response.php:366
Stack trace:
#0 /home/webmaster/web/live/public_html/vendor/yiisoft/yii2/web/Response.php(339): yii\web\Response->sendHeaders()
#1 /home/webmaster/web/live/public_html/vendor/yiisoft/yii2/web/ErrorHandler.php(136): yii\web\Response->send()
#2 /home/webmaster/web/live/public_html/vendor/craftcms/cms/src/web/ErrorHandler.php(139): yii\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#3 /home/webmaster/web/live/public_html/vendor/yiisoft/yii2/base/ErrorHandler.php(135): craft\web\ErrorHandler->renderException(Object(yii\base\ErrorException))
#4 /home/webmaster/web/live/public_html/vendor/craftcms/cms/src/web/ErrorHandler.php(63): yii\base\ErrorHandler->handleException(Object(yii\base\ErrorException))
#5 [internal function]: craft\web\ErrorHandler->handleException(Object(yii\base\ErrorException))
#6 {main}
Previous exception:
yii\base\ErrorException: ini_set(): A session is active. You cannot change the session module's ini settings at this time in /home/webmaster/web/live/public_html/web/index.php:28
Stack trace:
#0 /home/webmaster/web/live/public_html/vendor/craftcms/cms/src/web/ErrorHandler.php(76): yii\base\ErrorHandler->handleError(2, 'ini_set(): A se...', '/home/webmaster...', 28)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'ini_set(): A se...', '/home/webmaster...', 28, Array)
#2 /home/webmaster/web/live/public_html/web/index.php(28): ini_set('session.save_ha...', 'redis')
#3 {main}
```

**Comment by brandonkelly** [comment]
Looks like the offending `ini_set()` call is coming from line 28 of your `web/index.php` file.

**Comment by angrybrad** [comment]
@tonyclemmey any reason you can't use the https://github.com/yiisoft/yii2-redis library instead?

**Comment by tonyclemmey** [comment]
Hey, @angrybrad thanks for the follow-up. I did mention just above my `config/app.php` code the composer require command is run on deployment for yii2-redis.

Apologies, It is my first time configuring Craft 3 & Redis & in a Load Balanced Cluster, and the Craft docs as such I found very minimal so I was trying to follow the links above which mentioned the configuration lines to add in `web/index.php`, which was the cause of the internal server/session is already active issue.

It is strange on the same server ENV, it is possible to run a test PHP file with the same lines as to what I added in "web/index.php" and it will successfully use Redis for sessions instead of files /tmp, but the same does not work for `web/index.php`

It seems the only method of configuration for me with Craft at the moment is to just set the session handler and session path values in the PHP.ini file.

If you have any helpful info on what could be the reason why the values work in a test PHP outside Craft but `web/index.php` does not...it would be helpful to understand any idea.

Thanks again

**Comment by angrybrad** [comment]
@tonyclemmey Sorry, guess I should have read the whole ticket. :)

If you're using yii2-redis, then the only config you should need is in https://craftcms.com/docs/3.x/config/#redis-example and you shouldn't need to modify anything in your public index.php file.

**Comment by tonyclemmey** [comment]
@angrybrad That's the thing I did follow the Redis example you linked to, but still Craft would remain using files and /tmp instead of Redis. That's when I started browsing on Craft 3 Redis Config and came across the links I included, of which mention about setting the session handler and session path in `web/index.php`

Currently, if I do not set the session handler and session path direct in PHP.ini (master) Craft 3 does not use Redis, and if I set the values in `web/index.php` I get the internal server error / init_set() / session is already active.

I can completely appreciate that it could point towards server env configuration, but when using a test PHP file and setting the session handler and session path in the file, it is respected and Redis works.

So I must say I am not sure why it is different for Craft.

**Comment by angrybrad** [comment]
@tonyclemmey for saving PHP sessions in Redis, use this config: https://craftcms.com/docs/3.x/config/#redis-example-2

**Comment by tonyclemmey** [comment]
@angrybrad Yes if you look at my `config/app.php` code above I did in fact use the #redis-example-2 config you mentioned. I commented it out, as I also tested examples I found elsewhere, so I had 2x session components (switching between the 2) and the commented out getenv values just to test with hardcoded ones. 

The #redis-example-2 session config does not resolve REDIS being used for sessions unless I set the session handler and session path direct in PHP.ini (master) 

Which is not the case for a test PHP file which I can add the session handler and session path and its uses Redis. (No need to edit PHP.ini file)

The PHP.ini edit gets Craft 3 using Redis working for me now, but ideally, it would be prefered the site handle the setting of the session handler and session path.

If still, I can not get this configured Craft side, I will look at `.user.ini` file in the site dir.

Any more ideas or tips are very welcome as my first time trying to configure.
